plugins/redfish/fu-redfish-device.c

```diff
@@ -0,0 +1,739 @@
+		fu_device_add_instance_id (FU_DEVICE (self), devid1);
+	}
+	if (vendor_id != 0x0 && model_id != 0x0 &&
+	    subsystem_vendor_id != 0x0 && subsystem_model_id != 0x0) {
+		g_autofree gchar *devid2 = NULL;
+		devid2 = g_strdup_printf ("%s\\VEN_%04X&DEV_%04X&SUBSYS_%04X%04X",
+					  subsystem, (guint) vendor_id, (guint) model_id,
+					  (guint) subsystem_vendor_id,
+					  (guint) subsystem_model_id);
+		fu_device_add_instance_id (FU_DEVICE (self), devid2);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_redfish_device_probe_related_pcie_functions (FuRedfishDevice *self,
+						const gchar *uri,
+						GError **error)
+{
+	JsonObject *json_obj;
+	g_autoptr(FuRedfishRequest) request = fu_redfish_backend_request_new (self->backend);
+
+	/* get URI */
+	if (!fu_redfish_request_perform (request, uri,
+					 FU_REDFISH_REQUEST_PERFORM_FLAG_LOAD_JSON |
+					 FU_REDFISH_REQUEST_PERFORM_FLAG_USE_CACHE,
+					 error))
+		 return FALSE;
+	json_obj = fu_redfish_request_get_json_object (request);
+
+	if (json_object_has_member (json_obj, "Members")) {
+		JsonArray *members_array = json_object_get_array_member (json_obj, "Members");
+		for (guint i = 0; i < json_array_get_length (members_array); i++) {
+			JsonObject *related_item;
+			related_item = json_array_get_object_element (members_array, i);
+			if (json_object_has_member (related_item, "@odata.id")) {
+				const gchar *id = json_object_get_string_member (related_item, "@odata.id");
+				if (!fu_redfish_device_probe_related_pcie_item (self, id, error))
+					return FALSE;
+			}
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_redfish_device_probe_related_item (FuRedfishDevice *self,
+				      const gchar *uri,
+				      GError **error)
+{
+	JsonObject *json_obj;
+	g_autoptr(FuRedfishRequest) request = fu_redfish_backend_request_new (self->backend);
+
+	/* get URI */
+	if (!fu_redfish_request_perform (request, uri,
+					 FU_REDFISH_REQUEST_PERFORM_FLAG_LOAD_JSON |
+					 FU_REDFISH_REQUEST_PERFORM_FLAG_USE_CACHE,
+					 error))
+		 return FALSE;
+	json_obj = fu_redfish_request_get_json_object (request);
+
+	/* optional properties */
+	if (json_object_has_member (json_obj, "SerialNumber")) {
+		const gchar *tmp = json_object_get_string_member (json_obj, "SerialNumber");
+		if (tmp != NULL && tmp[0] != '\0' && g_strcmp0 (tmp, "N/A") != 0)
+			fu_device_set_serial (FU_DEVICE (self), tmp);
+	}
+
+	/* sometimes an array, sometimes an object! */
+	if (json_object_has_member (json_obj, "PCIeFunctions")) {
+		JsonNode *pcie_functions = json_object_get_member (json_obj, "PCIeFunctions");
+		if (JSON_NODE_HOLDS_OBJECT (pcie_functions)) {
+			JsonObject *obj = json_node_get_object (pcie_functions);
+			if (json_object_has_member (obj, "@odata.id")) {
+				const gchar *id = json_object_get_string_member (obj, "@odata.id");
+				if (!fu_redfish_device_probe_related_pcie_functions (self, id, error))
+					return FALSE;
+			}
```

Comment:
this seems like a duplicate block of code -> static function?

Reply:
Nah, it's subtly different unfortunately. I might make some convenience helpers for getting arrays of @odata.id objects if you think this is a good idea.